posix-list-dir_test.go

```diff
@@ -0,0 +1,137 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"testing"
+)
+
+func TestReadDir0(t *testing.T) {
+	// Check non existent directory.
+	if _, err := readDir("/tmp/non-existent-directory"); err != errFileNotFound {
+		t.Fatalf("expected = %s, got: %s", errFileNotFound, err)
+	}
+
+	// Check if file is given.
+	if _, err := readDir("/etc/issue/mydir"); err != errFileNotFound {
+		t.Fatalf("expected = %s, got: %s", errFileNotFound, err)
+	}
+
+	// Check if permission denied.
+	if _, err := readDir("/proc/1/fd"); err == nil {
+		t.Fatalf("expected = an error, got: nil")
+	}
+}
+
+type result struct {
+	dir     string
+	entries []string
+}
+
+var testResults = []result{}
+
+// Test to read empty directory.
+func setupTestReadDir1(t *testing.T) {
+	if dir, err := ioutil.TempDir("", "minio-posix-list-dir"); err != nil {
+		t.Fatal("failed to setup.", err)
+	} else {
+		testResults = append(testResults, result{dir, []string{}})
+	}
+}
+
+// Test to read empty directory with reserved names.
+func setupTestReadDir2(t *testing.T) {
+	if dir, err := ioutil.TempDir("", "minio-posix-list-dir"); err != nil {
+		t.Fatal("failed to setup.", err)
+	} else {
+		entries := []string{}
+		for _, reservedName := range posixReservedPrefix {
+			if err := ioutil.WriteFile(filepath.Join(dir, reservedName), []byte{}, os.ModePerm); err != nil {
+				sort.Strings(entries)
+				testResults = append(testResults, result{dir, entries})
+				t.Fatal("failed to setup.", err)
+			}
+		}
+
+		sort.Strings(entries)
+		testResults = append(testResults, result{dir, entries})
+	}
+}
+
+// Test to read non-empty directory.
+func setupTestReadDir3(t *testing.T) {
+	if dir, err := ioutil.TempDir("", "minio-posix-list-dir"); err != nil {
+		t.Fatal("failed to setup.", err)
+	} else {
+		entries := []string{}
+		for i := 0; i < 10; i++ {
+			name := fmt.Sprintf("file-%d", i)
+			entries = append(entries, name)
+			if err := ioutil.WriteFile(filepath.Join(dir, name), []byte{}, os.ModePerm); err != nil {
+				sort.Strings(entries)
+				testResults = append(testResults, result{dir, entries})
+				t.Fatal("failed to setup.", err)
+			}
+		}
+
+		sort.Strings(entries)
+		testResults = append(testResults, result{dir, entries})
+	}
+}
+
+func teardown() {
+	for _, r := range testResults {
+		os.RemoveAll(r.dir)
+	}
+}
+
+func checkResult(expected []string, got []string) bool {
+	if len(expected) != len(got) {
+		return false
+	}
+
+	for i := range expected {
+		if expected[i] != got[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func TestReadDir(t *testing.T) {
+	defer teardown()
+	setupTestReadDir1(t)
```

Comment:
The setup data can collected into cases like these https://github.com/minio/minio/blob/master/object-api-multipart_test.go#L1158 .